move amazon provider to own repository

durandom opened the pull request:
* removes travis test-suite
* links to https://github.com/durandom/manageiq-providers-amazon
* necessary changes to support running manageiq as a dummy test app in an engine

- [x] run travis test suite in amazon gem

@miq-bot add_labels wip, providers/amazon, darga/no

durandom commented:
Foreman runs tests of plugins (which are also created as rails engines) inside the whole test suite.
Which also makes sense.

What happens if some commit to miq makes the amazon gem tests fail?
We would only see that, once we commit something to the gem and run the test suite there.

For now, I guess we try this and see how if and how it messes testing up.

Down the road we could solve this by having a versioned and stable platform gem, that the provider gems depend upon.

Just mentioning that here, to have everybody on the same page.

see: http://projects.theforeman.org/projects/foreman/wiki/How_to_Create_a_Plugin#Testing

Fryguy commented:
@durandom I had started this locally already if you want to pair.

durandom commented:
> @durandom I had started this locally already if you want to pair.

Would :heart: to. I was thinking of adding ManageIQ/manageiq as a Gemfile dependency (which would need a .gemspec in ManageIQ/manageiq). This way bundler could take care of cloning. But then manageiq is not really a gem.

Another approach would be to shallow clone into spec/dummy. The upside would be: We have the manageiq sources right there when developing.

durandom commented:
@Fryguy check out https://github.com/durandom/manageiq-providers-amazon and the latest commit there to see what it took to make the aws tests run on travis. If you click on branches in this repo you have a link to the travis build

One thing I'm not satisfied with, is the way I handle dependencies. I copied all required dependencies into aws Gemfile to make the tests run. At least this is less than requiring everything from manageiq/Gemfile - and the list will shorten once we moved more providers into gems. But it duplicates the deps and they will run out of sync
We could move the overlapping stuff into a new Gemfile and require that inside the provider.

durandom commented:
To test this:

```shel
git clone git@github.com:durandom/manageiq-providers-amazon
cd manageiq-providers-amazon
# clone manageiq into spec/manageiq
bin/setup
# run tests
bin/rails app:test:manageiq-providers-amazon:setup app:test:manageiq-providers-amazon
```

Fryguy commented on Gemfile:
```diff
@@ -14,7 +14,7 @@ gem "deep_merge",                      "~>1.0.1", :git => "git://github.com/Mana
 path "gems/" do
   gem "manageiq_foreman", :require => false
-  gem "manageiq-providers-amazon"
+  gem "manageiq-providers-amazon", :git => "git://github.com/durandom/manageiq-providers-amazon", :branch => "master"
```
If this is git based it should move out of the enclosing `path` section.

Fryguy commented on lib/tasks/evm_test_helper.rb:
```diff
@@ -6,7 +6,10 @@ module EvmTestHelper
   def self.init_rspec_task(t, rspec_opts = [])
-    rspec_opts = ['--options', "\"#{Rails.root.join(".rspec_ci")}\""] + rspec_opts if ENV['CI']
+    if ENV['CI']
+      rspec_ci = defined?(ENGINE_ROOT) ? ENGINE_ROOT+"/.rspec_ci" : Rails.root.join(".rspec_ci")
```
Minor but prefer File.join or Pathname#join over `+`

durandom commented:
@miq-bot remove_label wip

Fryguy commented:
With @durandom , created a new official repo https://github.com/ManageIQ/manageiq-providers-amazon .  cc @blomquisg @chessbyte .

Additionally, I found out about a new feature on Travis since about May: https://docs.travis-ci.com/user/cron-jobs/ .  I've already requested permissions for that, so we can use it to run the plugin tests nightly.  This will allow us to catch failures on the plugin side that may have been caused by changes in ManageIQ proper.

chessbyte commented:
@durandom @Fryguy Should the `aws-sdk` gem be removed from Gemfile, now that it is required by the new repo?

Fryguy commented:
Yes, I believe so.

durandom commented:
> @durandom @Fryguy Should the aws-sdk gem be removed from Gemfile, now that it is required by the new repo?

nope, authentication/amazon still depends on it.

We could move that to the providers-amazon gem though, as I guess that anyone who wants to auth via aws also has it as a provider?!

Fryguy commented:
Yeah eventually, auth should probably made pluggable as well so that we can have other pluggable auths (like Keystone for OpenStack).

Fryguy commented:
Another place that needs to be `Rails.root`ed: https://github.com/ManageIQ/manageiq/blob/master/spec/spec_helper.rb#L39-L42

durandom commented:
> Another place that needs to be Rails.rooted: https://github.com/ManageIQ/manageiq/blob/master/spec/spec_helper.rb#L39-L42

:+1: fixed it and rebased again. hope to get this in, before we get merge conflicts again.
Will move the commit that caused the conflict to the new repo

durandom commented:
@Fryguy the travis failure is, because https://github.com/ManageIQ/manageiq-providers-amazon/pull/7 has to be merged first...

Fryguy commented:
Ok I merged the other PR and restarted the build here

durandom commented:
:tada: awesome...

Fryguy commented:
Great work @durandom !!! 🎉 🎈 🍰 

cc @blomquisg @jrafanie
